src/lib.rs

```diff
@@ -0,0 +1,534 @@
+//       It is unclear whether testing unlocked memory is something useful
+// TODO: Check for timeout, decrementing memory size can take non trivial time
+fn memory_resize_and_lock(
+    memory: &mut [usize],
+    allow_mem_resize: bool,
+) -> anyhow::Result<&mut [usize]> {
+    #[cfg(windows)]
+    {
+        crate::windows::memory_resize_and_lock(memory, allow_mem_resize)
+    }
+
+    #[cfg(unix)]
+    {
+        crate::unix::memory_resize_and_lock(memory, allow_mem_resize)
+    }
+}
+
+fn memory_unlock(memory: &mut [usize]) -> anyhow::Result<()> {
+    #[cfg(windows)]
+    {
+        crate::windows::memory_unlock(memory)
+    }
+
+    #[cfg(unix)]
+    {
+        crate::unix::memory_unlock(memory)
+    }
+}
+
+#[cfg(windows)]
+mod windows {
+    use {
+        crate::prelude::*,
+        std::mem::MaybeUninit,
+        windows::Win32::{
+            Foundation::ERROR_WORKING_SET_QUOTA,
+            System::{
+                Memory::{VirtualLock, VirtualUnlock},
+                SystemInformation::{GetNativeSystemInfo, SYSTEM_INFO},
+                Threading::{
+                    GetCurrentProcess, GetProcessWorkingSetSize, SetProcessWorkingSetSize,
+                },
+            },
+        },
+    };
+
+    // TODO: We don't really need to replace the set size if if it is much larger than memsize?
+    pub(super) fn replace_set_size(memsize: usize) -> anyhow::Result<(usize, usize)> {
+        let (mut min_set_size, mut max_set_size) = (0, 0);
+        unsafe {
+            GetProcessWorkingSetSize(GetCurrentProcess(), &mut min_set_size, &mut max_set_size)
+                .context("Failed to get process working set")?;
+            // TODO: Not sure what the best choice of min and max should be
+            SetProcessWorkingSetSize(
+                GetCurrentProcess(),
+                memsize.saturating_mul(2),
+                memsize.saturating_mul(4),
+            )
+            .context("Failed to set process working set")?;
+        }
+        Ok((min_set_size, max_set_size))
+    }
+
+    pub(super) fn restore_set_size(min_set_size: usize, max_set_size: usize) -> anyhow::Result<()> {
+        unsafe {
+            SetProcessWorkingSetSize(GetCurrentProcess(), min_set_size, max_set_size)
+                .context("Failed to set process working set")
+        }
+    }
+
+    // TODO: Resize according to min set size instead of decrementing
+    pub(super) fn memory_resize_and_lock<'a>(
+        mut memory: &'a mut [usize],
+        allow_mem_resize: bool,
+    ) -> anyhow::Result<&'a mut [usize]> {
+        let page_size = unsafe {
+            let mut sysinfo: MaybeUninit<SYSTEM_INFO> = MaybeUninit::uninit();
+            GetNativeSystemInfo(sysinfo.as_mut_ptr());
+            let sysinfo = sysinfo.assume_init();
+            usize::try_from(sysinfo.dwPageSize).context("Failed to convert page size to usize")?
+        };
+
+        loop {
+            unsafe {
```

Comment:
Just wrap the `unsafe {}` around the `VirtualLock()` call. Everything else this code does is safe.